apps/stacker/app.js

```diff
@@ -0,0 +1,140 @@
+const TICKRATE = 80;
+const BLOCK_SIZE = 16;
+const BLOCK_COLOR_PRIMARY = "#ff0000";
+const BLOCK_COLOR_SECONDARY = "#ffff00";
+const GAMEBOARD_X = 16;
+const GAMEBOARD_WIDTH = g.getWidth() - 16 - BLOCK_SIZE;
+const START_X = GAMEBOARD_X;
+const START_Y = g.getHeight() - BLOCK_SIZE - 1;
+const START_LENGTH = 5;
+var startTime = 0;
+var length = 5;
+var updateTimeout;
+var rows = [];
+var playing = true;
+
+class Block {
+  constructor(x, y, match) {
+    this.x = x;
+    this.y = y;
+    this.match = match;
+    this.show = true;
+  }
+}
+
+class Row {
+ constructor(x, y, size, direction, match) {
+   this.y = y;
+   this.size = size;
+   this.blocks = [];
+   if (Math.random() > 0.49) {
+     this.direction = 1;
+     this.x = BLOCK_SIZE;
+   }
+   else {
+     this.direction = -1;
+     this.x = g.getWidth() - this.size * BLOCK_SIZE;
+   }
+   this.match = match;
+   for (var i = 0; i < size; i++) {
+    var b = new Block(this.x + (BLOCK_SIZE * i), this.y, this.match);
+    this.blocks.push(b);
+  }
+ }
+  update() {
+    this.x += BLOCK_SIZE * this.direction;
+    if (this.x + (this.size * BLOCK_SIZE) > GAMEBOARD_X + GAMEBOARD_WIDTH || this.x < GAMEBOARD_X) {
+      this.direction = -this.direction;
+    }
+    for (var i = 0; i < this.size; i++) {
+      this.blocks[i].x = this.x + BLOCK_SIZE * i;
+    }
+  }
+  draw() {
+    for (var i = 0; i < this.size; i++) {
+      if (this.blocks[i].show) {
+        g.drawRect({x: this.blocks[i].x, y: this.y, w: BLOCK_SIZE, h: BLOCK_SIZE});
+      }
+    }
+  }
+}
+
+function init() {
```

Comment:
You can add touch screen functionality that works just like the button presses and reduce some of the same code by adding a `gameOver` function like this:
```
function init() {
  Bangle.setLCDPower(1);
  g.setTheme({bg:"#000", fg:"#fff", dark:true}).clear();
  setInterval(update, TICKRATE);
  setWatch(input_pressed, BUTTON);
  Bangle.on('touch', input_pressed); // Add initial touch listener
  var test = new Row(START_X, START_Y, length, 1, true);
  rows.push(test);
}

function update() {
  if (playing) {
    g.clear(reset);
    rows[rows.length - 1].update();
    rows.forEach(row => row.draw());
  }
}

function resetGame() {
  playing = true;
  rows = [];
  length = START_LENGTH;
  var test = new Row(START_X, START_Y, length, 1, true);
  rows.push(test);
  update();
  Bangle.removeListener('touch', resetGame); // Remove reset game touch listener
  Bangle.on('touch', input_pressed); // Switch back to input touch listener
}

function gameOver(message) {
  print(message);
  g.clear(reset);
  E.showMessage(message);
  playing = false;
  setWatch(resetGame, BUTTON, {repeat:0, debounce:50, edge:"rising"});
  Bangle.removeListener('touch', input_pressed); // Remove input touch listener
  Bangle.on('touch', resetGame); // Add touch listener to reset the game
}

function lose() {
  gameOver("YOU LOSE!");
}

function win() {
  gameOver("YOU WIN!");
}
```